rasa/core/lock_store.py

```diff
@@ -329,3 +354,155 @@ def _load_from_module_name_in_endpoint_config(
+                in case Redis doesn't respond within `socket_timeout` seconds.
+        """
+        import redis
+
+        self.red = redis.StrictRedis(
+            host=host,
+            port=int(port),
+            db=int(db),
+            password=password,
+            ssl=use_ssl,
+            socket_timeout=socket_timeout,
+        )
+
+        self.key_prefix = DEFAULT_CONCURRENT_REDIS_LOCK_STORE_KEY_PREFIX
+        if key_prefix:
+            logger.debug(f"Setting non-default redis key prefix: '{key_prefix}'.")
+            self._set_key_prefix(key_prefix)
+
+        super().__init__()
+
+    def _set_key_prefix(self, key_prefix: Text) -> None:
+        if isinstance(key_prefix, str) and key_prefix.isalnum():
+            self.key_prefix = (
+                key_prefix + ":" + DEFAULT_CONCURRENT_REDIS_LOCK_STORE_KEY_PREFIX
+            )
+        else:
+            logger.warning(
+                f"Omitting provided non-alphanumeric redis key prefix: '{key_prefix}'. "
+                f"Using default '{self.key_prefix}' instead."
+            )
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Retrieves lock (see parent docstring for more information)."""
+        tickets: Deque[Ticket] = deque()
+
+        pattern = self.key_prefix + conversation_id + ":" + "[0-9]*"
+        redis_keys = self.red.keys(pattern)
+
+        for key in redis_keys:
+            serialised_ticket = self.red.get(key)
+            ticket = Ticket.from_dict(json.loads(serialised_ticket))
+            tickets.appendleft(ticket)
+
+        tickets = deque(sorted(tickets, key=lambda x: x.number))
+
+        return TicketLock(conversation_id, tickets)
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Deletes lock for conversation ID."""
+        pattern = self.key_prefix + conversation_id + ":*"
+        redis_keys = self.red.keys(pattern)
+
+        if not redis_keys:
+            logger.debug(
+                f"The lock store does not contain any key-value "
+                f"items for conversation '{conversation_id}'."
+            )
+            return None
+
+        deletion_successful = self.red.delete(*redis_keys)
+        if deletion_successful == 0:
+            self._log_deletion(conversation_id, False)
+        else:
+            self._log_deletion(conversation_id, True)
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit individual tickets and last issued ticket number to storage."""
+        last_issued_ticket = lock.tickets[-1]
+        serialised_ticket = last_issued_ticket.dumps()
+        key = (
+            self.key_prefix
+            + lock.conversation_id
+            + ":"
+            + str(last_issued_ticket.number)
+        )
+        self.red.set(
+            name=key, value=serialised_ticket, ex=int(last_issued_ticket.expires)
+        )
+
+        # set expiry on the last_issued_ticket_number key too
+        last_issued_key = (
```

Comment:
This was added because the `test_concurrent_lock_expiration` was failing at the last assertion line, where it expects to get the ticket number that an expired ticket had. 
If we don't want this mechanism, as we discussed in the other thread about `finish_serving` method, I can remove this code and amend the test accordingly. And I will add in the docs PR that I will prepare a mention that ticket numbers are unique to ticket instances, so a ticket number will not be reassigned if the ticket expires. Let me know if that sounds good.

Reply:
Done: https://github.com/RasaHQ/rasa/pull/11256/commits/5b7e619f217f4b5973c37814fc321222ebd7abc8

Reply:
sounds good :+1: